tests/test_data_utils/test_downsample_cells.py

```diff
@@ -1,98 +1,150 @@
+                self.df_multi_random,
+                ['nonexistent_column', 'region']
+            )
 
     def test_downsample_without_stratify(self):
         """Downsample without stratify"""
         n_samples = 20
-        df_downsampled = downsample_cells(
-            self.df, self.annotation,
+        df_downsampled_single = downsample_cells(
+            self.df_single_random, self.annotation_single,
+            n_samples=n_samples, stratify=False
+        )
+        df_downsampled_multi = downsample_cells(
+            self.df_multi_random, self.annotation_multi,
             n_samples=n_samples, stratify=False
         )
         self.assertTrue(
-            all(df_downsampled.groupby(
-                self.annotation
+            all(df_downsampled_single.groupby(
+                self.annotation_single
             ).size().values <= n_samples)
         )
-
-    def test_downsample_with_stratify_without_rand(self):
-        """Downsample with stratify without random"""
-        n_samples = 20
-        df_downsampled = downsample_cells(
-            self.df, self.annotation,
-            n_samples=n_samples, stratify=True
+        # For multi obs, group by both columns without combining them
+        self.assertTrue(
+            all(df_downsampled_multi.groupby(
+                self.annotation_multi
+            ).size().values <= n_samples)
         )
-        self.assertEqual(df_downsampled.shape[0], n_samples)
 
-    def test_downsample_with_stratify_with_rand(self):
-        """Downsample with stratify with random"""
-        n_samples = 20
-        df_downsampled = downsample_cells(
-            self.df, self.annotation,
-            n_samples=n_samples, stratify=True, rand=True
-        )
-        self.assertEqual(df_downsampled.shape[0], n_samples)
+    def check_stratified_counts(
+            self, downsampled_df, columns, expected_counts):
+        """
+        Helper method to check stratified counts for both single and
+        multi annotations.
+        """
+        combined_col = '_'.join(columns)
 
-    def test_downsample_no_n_samples(self):
-        """Downsample without n_samples should return the original dataframe"""
-        df_downsampled = downsample_cells(self.df, self.annotation)
-        pd.testing.assert_frame_equal(self.df, df_downsampled)
+        # Ensure the combined column exists in the downsampled_df
+        if combined_col not in downsampled_df.columns:
+            downsampled_df[combined_col] = downsampled_df[columns].apply(
+                lambda row: '_'.join(row.values.astype(str)), axis=1
+            )
 
-    def test_stratification_frequency(self):
-        """Test the frequency of stratification"""
-        n_samples = 10
+        actual_counts = downsampled_df[combined_col].value_counts()
+
+        for key, expected_count in expected_counts.items():
+            self.assertEqual(expected_count, actual_counts.get(key, 0))
+
+        # Remove the combined column after checks
```

Comment:
@fangliu117, I am not sure why this code is in the unit test?

Reply:
Actually this was a helper function to define the expect counts. However, in the new commit, we include the manually counted true numbers of expected count, therefore this function was removed.